### PR TITLE
Detect encoding problems

### PR DESCRIPTION
**Pull request by ben-Draeger**
SDCcc currently accepts the encoding specified in a message and uses it to decode the message. When the encoding is specified incorrectly or the message contains characters that are not mappable using the specified encoding, then they are replaced by Charset-specific default characters, which may lead to False-positives or False-negatives in Test Cases that rely on messages stored in the database.

A much better approach would be to check the encoding while decoding the messages and to raise an Encoding Error when decoding a message with the specified encoding fails. As usual, these Encoding Errors can be managed using the EnableMessageEncodingCheck and SummarizeMessageEncodingErrors configuration options.

# Checklist

The following aspects have been respected by the author of this pull request, confirmed by both pull request assignee **and** reviewer:

* Adherence to coding conventions
  * [x] Pull Request Assignee
  * [x] Reviewer
* Adherence to javadoc conventions
  * [x] Pull Request Assignee
  * [x] Reviewer
* Changelog update (necessity checked and entry added or not added respectively)
  * [x] Pull Request Assignee
  * [x] Reviewer
* README update (necessity checked and entry added or not added respectively)
  * [x] Pull Request Assignee
  * [x] Reviewer
* config update (necessity checked and entry added or not added respectively)
  * [x] Pull Request Assignee
  * [x] Reviewer
* SDCcc executable ran against a test device (if necessary)
  * [x] Pull Request Assignee
  * [x] Reviewer

### CODE REVIEW

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java [hunk 1]
```diff
@@ -1864,8 +1867,7 @@ public void testDetermineCharsetFromMessageFromHttpHeader(@TempDir final File di
 
             // then
             assertEquals(Charset.forName("ISO-8859-13"), actualCharset);
-            Mockito.verify(this.testRunObserver, VerificationModeFactory.atLeastOnce())
-                    .invalidateTestRun(anyString());
+            verify(this.testRunObserver, VerificationModeFactory.atLeastOnce()).invalidateTestRun(anyString());
```
Comment (maximilianpilz) [inline sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:1870]: what is the purpose of this change ? 

seems unnecessary

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java [hunk 2]
```diff
@@ -2521,10 +2508,280 @@ public void testDetermineCharsetFromMessageASCIISpecialCase(@TempDir final File
...
+            final var result = messageStorage.convertMessageToMessageContent(message);
+
+            // then
+            assertNotNull(result);
+            assertFalse(testRunObserver.isInvalid());
+        }
+    }
+
+    /**
+     * Checks that convertMessageToMessageContent() works when it does not detect any decoding problems in the message.
```
Comment (maximilianpilz) [inline sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2558]: this is equal to the description for testConvertToMessageContent

has to be changed, since it does not fit here

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java [hunk 3]
```diff
@@ -2521,10 +2508,280 @@ public void testDetermineCharsetFromMessageASCIISpecialCase(@TempDir final File
...
+            // when
+            final var result = messageStorage.convertMessageToMessageContent(message);
+            // then
+            assertNotNull(result);
+        }
+        verify(testRunObserver, times(4)).invalidateTestRun(anyString());
+    }
+
+    /**
+     * Checks that convertMessageToMessageContent() works when it does not detect any decoding problems in the message.
```
Comment (maximilianpilz) [inline sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2599]: this is equal to the description for testConvertToMessageContent

has to be changed, since it does not fit here

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java [hunk 4]
```diff
@@ -2521,10 +2508,280 @@ public void testDetermineCharsetFromMessageASCIISpecialCase(@TempDir final File
...
+            final var result = messageStorage.convertMessageToMessageContent(message);
+
+            // then
+            assertNotNull(result);
+        }
+        verify(testRunObserver, times(6)).invalidateTestRun(anyString());
+    }
+
+    /**
+     * Checks that convertMessageToMessageContent() works when it does not detect any decoding problems in the message.
```
Comment (maximilianpilz) [inline sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2642]: this is equal to the description for testConvertToMessageContent

has to be changed, since it does not fit here

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java [hunk 5]
```diff
@@ -2521,10 +2508,280 @@ public void testDetermineCharsetFromMessageASCIISpecialCase(@TempDir final File
...
+     * @param dir - a temporary directory.
+     * @throws IOException - when something goes wrong.
+     */
+    @Test
+    public void testConvertToMessageContent(@TempDir final File dir) throws IOException {
+
+        final Charset charsetInHttpHeader = StandardCharsets.UTF_8;
+        final Charset charsetInXMLDeclaration = StandardCharsets.UTF_8;
+        final Charset charsetInXMLDeclarationEncoding = StandardCharsets.UTF_8;
```
Comment (maximilianpilz) [inline sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2524]: I suggest to change to an encoding other than the default/fallback, at best one that would throw an error with the default.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java [hunk 6]
```diff
@@ -2521,10 +2508,280 @@ public void testDetermineCharsetFromMessageASCIISpecialCase(@TempDir final File
...
+            final var result = messageStorage.convertMessageToMessageContent(message);
+
+            // then
+            assertNotNull(result);
+        }
+        verifyNoInteractions(testRunObserver);
+    }
+
+    /**
+     * Checks that determineCharsetFromMessage() works with the given Charsets.
```
Comment (maximilianpilz) [inline sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2777]: please reformat the multiline comment, so that different params are aligned

for comments of other methods as well please

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java [hunk 7]
```diff
@@ -2521,10 +2508,280 @@ public void testDetermineCharsetFromMessageASCIISpecialCase(@TempDir final File
...
+                    "<?xml version=\"1.0\" encoding=\"%s\"?>%n<sometag></sometag>", charsetInXMLDeclaration);
+            final byte[] encodedContent = content.getBytes(charsetInXMLDeclarationEncoding);
+            message.write(encodedContent, 0, encodedContent.length);
+            message.close();
+
+            // when
+            final var result = messageStorage.convertMessageToMessageContent(message);
+
+            // then
+            assertNotNull(result);
```
Comment (maximilianpilz) [inline sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2552]: verify the content of the body field too

**Review by maximilianpilz — CHANGES_REQUESTED** [review]

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 1 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:1870]: Agreed, it is not necessary.
However, removing redundant information to make the code more readable is still best practice.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 5 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2524]: Unfortunately, using anything but UTF-8 here causes the TestRun to be invalidates due to MDPWS:R0007_0 and would hence not constitute a good case.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 7 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2552]: Done.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 4 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java]: Done.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 3 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java]: Done.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 2 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java]: Done.

File: CHANGELOG.md [hunk 8]
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   SDCcc.SummarizeMessageEncodingErrors.
 - checking for message encoding errors can be disabled using the configuration option SDCcc.EnableMessageEncodingCheck.
 - configuration option SDCcc.Network.MulticastTTL to configure the Time To Live of the Multicast Packets used for Discovery.
+- SDCcc now checks the encodings of all messages when SDCcc.EnableMessageEncodingCheck is enabled.
```
[the remark this answers is not in the document]
Comment (ben-Draeger) [reply CHANGELOG.md]: Done.

File: sdccc/src/main/java/com/draeger/medical/sdccc/messages/MessageStorage.java [hunk 9]
```diff
@@ -270,8 +274,27 @@ private MessageContent convertMessageToMessageContent(final Message message) {
...
+            charsetDecoder.onUnmappableCharacter(CodingErrorAction.REPORT);
+            charsetDecoder.onMalformedInput(CodingErrorAction.REPORT);
+            try {
+                final var decodingResult = charsetDecoder.decode(ByteBuffer.wrap(message.getFinalMemory()));
+                body = decodingResult.toString();
+            } catch (CharacterCodingException e) {
+                if (this.enableEncodingCheck) {
+                    if (this.summarizeMessageEncodingErrors) {
+                        // TestRun will be invalidated in TestSuite if messageEncodingErrorCount > 0
+                        this.messageEncodingErrorCount += 1;
```
[the remark this answers is not in the document]
Comment (ben-Draeger) [reply sdccc/src/main/java/com/draeger/medical/sdccc/messages/MessageStorage.java]: Done.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 6 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2804]: Done.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 5 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2545]: I tried creating another test case to cover the case, but found that it is impossible. The case is impossible to reach and consequently there is no control flow path for that case that could be tested.

File: CHANGELOG.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ben-Draeger) [reply CHANGELOG.md]: This formulation has been completely redone. Please check if the reader is given a precise enough impression of what is really going on.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 1 above)
Comment (maximilianpilz) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:1871]: How did it become more readable? If anything, it contains less information now.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 5 above)
Comment (maximilianpilz) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2545]: > The case is impossible to reach and consequently there is no control flow path for that case that could be tested.

If that would be true, the mentioned part of the code would be unnecessary and could be removed.
However I do not agree with this assessment, it is reachable as some of the already exisiting unit tests proof.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 1 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:1871]: The information that "verify" is a static method from Mockito is redundant (it is also contained in the imports). Also, it is not neccessary as verify is not ambiguous.
Removing it at this place does hence not decrease the amount of information.
On the contrary, one has to read less code to gain all the relevant information, which is why I call it "more readable".

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 5 above)
Comment (ben-Draeger) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:2546]: Misunderstanding resolved, Test case added.

File: sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java (hunk 1 above)
Comment (maximilianpilz) [reply sdccc/src/test/java/com/draeger/medical/sdccc/messages/TestMessageStorage.java:1872]: Well, I do not agree with you, but I will resolve this conversation nontheless, because it is not important enough to let it block the pr.

**Review by maximilianpilz — APPROVED** [review]
